File: leetcode/company-wise/bloomberg/0716_max_stack.cpp
// LC 716. Max Stack | Hard | Bloomberg
#include <list>
#include <map>
#include <stack>
using namespace std;

class MaxStack {
    list<int> dll;
    map<int, vector<list<int>::iterator>> mp;
public:
    void push(int x) { dll.push_back(x); mp[x].push_back(prev(dll.end())); }
    int pop() {
        int val = dll.back(); dll.pop_back();
        mp[val].pop_back(); if (mp[val].empty()) mp.erase(val);
        return val;
    }
    int top() { return dll.back(); }
    int peekMax() { return mp.rbegin()->first; }
    int popMax() {
        int val = mp.rbegin()->first;
        auto it = mp[val].back(); mp[val].pop_back();
        if (mp[val].empty()) mp.erase(val);
        dll.erase(it); return val;
    }
};

File: leetcode/company-wise/bloomberg/0716_max_stack.py
# LC 716. Max Stack | Hard | Bloomberg
from sortedcontainers import SortedList

class MaxStack:
    def __init__(self):
        self.stack = SortedList(key=lambda x: x[1])
        self.vals = SortedList()
        self.cnt = 0

    def push(self, x):
        self.stack.add((self.cnt, x)); self.vals.add((x, self.cnt)); self.cnt += 1

    def pop(self):
        idx, val = self.stack.pop()
        self.vals.remove((val, idx)); return val

    def top(self): return self.stack[-1][1]

    def peekMax(self): return self.vals[-1][0]

    def popMax(self):
        val, idx = self.vals.pop()
        self.stack.remove((idx, val)); return val
